ToolChain/Assembler/Assembler.py

```diff
@@ -56,79 +58,109 @@ def __init__(self, inputFile=None, outputFile=None):
+
+        for line in fileLines:
+            lineno += 1
+            # Here we begin our evaluation of each line. We get the binary instruction, offset, and label flag
+            # Memory locations are relative to the beginning of the file. Instructions and data are given set lengths.
+            # Internal and external symbol lists contain tuples containing the symbol name, and its offset.
+            # These will be used to fill the .o file information before the binary data.
+
+            try:
+                # This will catch any exception thrown by the parser. If there is one, we print the error,
+                # The line at which it occurred, and we quit out.
+                instruction, offset, labelFlag = parser.parse(line)
+                relativeAddressCounter += offset
+            except ValueError as e:
+                print(str(e) + " at line: {}".format(lineno))
+                quit()
+
+            if labelFlag == INSTRUCTION_FLAG:
+                # If the label flag is set to 0, it's a regular instruction, string, number, of memref.
+                binaryDataString += instruction
+
+            elif labelFlag == LOCAL_REFERENCE_FLAG:
+                # If the label flag is set to 1, it's an internal symbol. We ensure labels are unique when declared
+                if instruction not in localSymbols:
+                    localSymbols.append(instruction)
+                    labelTuple = (instruction, relativeAddressCounter)
+                    internalSymbols.append(labelTuple)
+                else:
+                    raise ValueError("Label already present in local list, cannot declare duplicate label name")
+                if instruction in globalSymbols:
+                    externalSymbols.append(labelTuple)
+
+            elif labelFlag == GLOBAL_REFERENCE_FLAG:
+                # If the label flag is set to 2, it's an external symbol. We ensure labels are unique when declared
+                if instruction not in globalSymbols:
+                    if instruction not in localSymbols:
+                        globalSymbols.append(instruction)
+                    else:
+                        # We can't have a global label declared if it has already been set as a local one
+                        raise ValueError("Label already present in local list, cannot be declared as global")
                 else:
-                    raise ValueError("Unresolved reference: {} is exported but "
-                                     "is never declared.".format(reference))
-        fileContent += b"</ExternalSymbols>"
+                    raise ValueError("Label already present in global list, cannot declare duplicate label name")
 
-        # Deal with the text section
-        fileContent += b"<Text>"
-        for instruction in self.parser.instructionList:
-            fileContent += instruction
-        fileContent += b"</Text>"
+            elif labelFlag == COMMENT_FLAG:
```

Comment:
Positioned at the end like this, I am questioning the usefulness of this block. Could it not be removed without causing issues?